src/beamlime/applications/base.py

```diff
@@ -215,14 +210,43 @@ def register_daemon(self, daemon: DaemonInterface) -> None:
         """
         self.daemons.append(daemon)
 
+    def cancel_all_tasks(self) -> None:
+        """Cancel all tasks."""
+        for task in self.tasks:
+            task.cancel()
+
+    @asynccontextmanager
+    async def _daemon_wrapper(
+        self, daemon: DaemonInterface
+    ) -> AsyncGenerator[None, None]:
+        try:
+            self.info('Running daemon %s', daemon.__class__.__qualname__)
+            yield
+        except Exception as e:
+            # Make sure all other async tasks are cancelled.
+            # It is because raising an exception will destroy only the task
+            # that had an error raised and may not affect other tasks in some cases,
+            # e.g. in Jupyter Notebooks.
+            self.error(f"Daemon {daemon} failed. Cancelling all other tasks...")
+            # Break all daemon generator loops.
+            self._break = True
+            # Let other daemons/handlers clean up.
+            await self.message_router.route(self.Stop(None))
+            # Make sure all other async tasks are cancelled.
+            self.cancel_all_tasks()
```

Comment:
Is there actually time for the handlers to stop? Or does this just cancel all tasks before they can be resumed and handle the stop message?

Or are message handlers and tasks completely separate?

Reply:
> Or are message handlers and tasks completely separate?

Yes, they are separate.
Even if ``message_router.run`` is canceled, since ``route`` is directly called here so it should be able to call all handlers that are subscribing ``Stop``.

Reply:
The coroutines (that are used to create the tasks) are dynamically created, using `DaemonInterface` run coroutines when `Application.run` is called.

And the handlers are called by ``MessageRouter.route``, which is constantly called in ``MessageRouter(DaemonInterface).run``.

So, if a ``MessageRouter.run`` task is canceled, even if ``Stop`` message is populated, it will not be routed automatically anymore.

But daemon objects including `MessageRouter` are still alive,  so we can still use ``route`` method  to route``Stop`` to handlers.

Did it answer your question....?